engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -174,12 +161,66 @@ public void shutdown() {
   }
 
   /**
-   * The worker action for the underlying logic of this module. At least one such thread must be
-   * spawned in order for this module to be operational.
+   * Processes {@link Item}s eligible for GC. Plays two roles. First of all cleans {@link
+   * #referenceQueue} in {@link #run()} method running in its own thread. Then it invokes finalizers
+   * in {@link #perform} method inside of Enso execution context.
    */
-  private class Runner implements Runnable {
+  private final class ProcessItems extends ThreadLocalAction implements Runnable {
+    /**
+     * @GuardedBy("pendingItems")
+     */
+    private final List<Item> pendingItems = new ArrayList<>();
+
     private volatile boolean killed = false;
 
+    ProcessItems() {
+      super(false, false, true);
+    }
+
+    /**
+     * Runs at a safe point in middle of regular Enso program execution. Gathers all available
+     * {@link #pendingItems} and runs their finalizers. Removes all processed items from {@link
+     * #pendingItems}. If there are any remaining, continues processing them. Otherwise finishes.
+     *
+     * @param access not used for anything
+     */
+    @Override
+    protected void perform(ThreadLocalAction.Access access) {
+      for (; ; ) {
+        Item[] toProcess;
+        synchronized (pendingItems) {
+          if (pendingItems.isEmpty() || pendingItems.get(0) == null) {
+            // nothing to process or already processing
+            // avoids re-entrant calls into this method
+            return;
+          }
+          toProcess = pendingItems.toArray(Item[]::new);
+          // mark as being processed
+          pendingItems.set(0, null);
```

Comment:
Why use 'magic value' `null` to mark `alreadyProcessing`? Cannot we use an explicit `boolean` flag instead? It would be much more clear what is happening IMO.

Reply:
I had `boolean` flag in one of previous versions.  However this is simpler.

> more clear what is happening

C'mon. Just read comment on line 193, 194 and 198 and you can figure it out!